### ISSUE TITLE
Move TDZ errors to no-undef?

### ISSUE THREAD

**Issue by jrvidal** [eslint/eslint#2877]
Example:
```js
x();

let x = () => {};
```
Currently, this triggers `no-use-before-define`. Wouldn't it make more sense to move TDZ warnings to another rule? I'd argue that `no-use-before-define` exists to disallow an error-prone pattern, such as:

```js
function foo() {
    return x;
}

var x = 1;

```

However, TDZ errors are static `ReferenceError`s, and constitute more than an error-prone pattern. I think it would make sense to report them as part of `no-undef`.

**Comment by nzakas** [comment]
I disagree. Looking at it from a user's perspective, it makes more sense to say "You used x before it was defined" vs. "x is not defined". 

The goal of `no-use-before-define` is to pretend that hoisting doesn't exist for functions and `var` variables, effectively mimicking block bindings, so I think it makes sense to keep the rule as-is.

**Comment by michaelficarra** [comment]
I agree with the OP here. @nzakas I don't understand your reasoning here at all.

> The goal of `no-use-before-define` is to pretend that hoisting doesn't exist for functions and `var` variables

Then why would it ever report TDZ errors?

**Comment by nzakas** [comment]
```js
foo();  //used before it was define

function foo(){};
```

and

```js
foo();   // used before it was defined

let foo = () =>{};
```

I think it makes sense to have the same error message in both cases.

**Comment by michaelficarra** [comment]
One will still work (so I'd configure it as a warning), and one will fail unconditionally when execution reaches that expression (so I'd configure it as an error). That's the difference.

**Comment by nzakas** [comment]
As I said, the goal is to mimic block bindings for `var`, so I think it makes sense to have actual block bindings handled the same way. We aren't flagging runtime errors, we are flagging code patterns that may or may not cause runtime errors.

**Comment by UltCombo** [comment]
@nzakas function declarations are available in the entire scope where they're defined. It is even a [common pattern](http://andrewkelley.me/post/js-callback-organization.html) to define functions below where they are called, as to maintain the reading flow uni-directional.

Also, consider this perfectly valid code:

```js
function a() {
	"use strict";
	if (Math.random() < .5) {
		b();
	}
}

function b() {
	"use strict";
	if (Math.random() < .5) {
		a();
	}
}
```

This will result in a "b was used before it was defined" linting error, which is impossible to solve.

**Comment by jrvidal** [comment]
My initial motivation to open this issue is the fact that TDZ errors seem to me a little bit more "erroneous" than hoisting footguns. I don't have a strong opinion against nzakas' point about whether it makes sense to keep everything under the same rule.

I think I personally would be content with just making `no-use-before-define` more "serious" about TDZ issues, meaning (a) an explicit mention in the rule docs and maybe (b) a separate error message. It seems to me that (a) is clearly beneficial on its own, so I'll open a PR for it.

**Comment by UltCombo** [comment]
@jrajav Thanks. :+1: I agree, TDZ means an actual (runtime) error, while simple ordering of function declarations is just a personal preference matter.

Note that catching all TDZ errors is pretty much impossible with just static analysis, I recall various discussions around this in Traceur and Babel. Consider these two examples:

```js
f();
let x;
function f() {
    x; // TDZ error, `x` is not initialized yet
}
```

```js
function f() {
    x; // works fine, the lexically scoped `x` binding has already been initialized when `f` executes
}
let x;
f();
```
